File: ceph-export.py
import os
import sys
import yaml
import json
import shutil
import argparse
import subprocess
import configparser

KEYRING_FILE = 'ceph.client.{}.keyring'


class Defaults(object):
    user = "admin"
    conf_dir = '/etc/ceph'
    output = '~/rhcs-config-export.yaml'
    output_format = 'yaml'


class Choices(object):
    output_format = ['yaml']


def ready():
    if not shutil.which('ceph'):
        return False, "ceph command not found!"
    if not os.path.isfile(os.path.join(args.confdir, 'ceph.conf')):
        return False, "ceph.conf not found"
    if not os.path.isfile(os.path.join(args.confdir, 
                          KEYRING_FILE.format(args.user))) and \
      not os.path.isfile(os.path.join(args.confdir, 'keyring-store', 'keyring')):
        return False, "missing keyring"
    return True, ""

def get_config(filename):
    conf = configparser.ConfigParser()
    conf.read(filename)
    return conf

def find_keyring():
    conf = None

    ceph_conf_keyring = os.path.join(args.confdir, KEYRING_FILE.format(args.user))
    ceph_conf_keystore = os.path.join(args.confdir, 'keyring-store', 'keyring')
    if os.path.isfile(ceph_conf_keyring):
        conf = get_config(ceph_conf_keyring)
    elif os.path.isfile(ceph_conf_keystore):
        conf = get_config(ceph_conf_keystore)

    if not conf or not conf.sections():
        # conf empty or can't read it!
        return None
    if 'client.admin' not in conf.sections():
        # no client.admin section
        return None
    if 'key' not in conf['client.' + args.user].keys():
        # no key in client.admin section
        return None

    return conf['client.' + args.user]['key']

def abort(message=None):
    error_str = ": {}".format(message) if message else ""
    print("Unable to continue{}".format(error_str))
    sys.exit(1)

def help():
    print("help me...")

def send_command(cmd):
    cmd_stream = cmd.split(' ')
    _c = subprocess.Popen(cmd_stream,
                          stdout=subprocess.PIPE,
                          stderr=subprocess.STDOUT)
    return _c.communicate()

def write_file(content):
    
    try:
        with open(os.path.expanduser(args.output), 'w') as f:
            f.write(content)
    except:
        print("Unexpected problem writing the file, dumping config here")
        print(output)
    else:
        print("Config written to {}".format(os.path.expanduser(args.output)))

def _dump_yaml(settings):
    
    output = yaml.safe_dump(settings, indent=2, explicit_start=2)
    write_file(output)

def dump(settings):
    if args.format == 'yaml':
        _dump_yaml(settings)
    else:
        abort("Unknown output format requested")    

def parse_args():
    parser = argparse.ArgumentParser(description="Export RHCS configuration information",
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("-o", "--output", type=str, required=False,
                        default=Defaults.output, help="output file for the export")
    parser.add_argument("-c", "--confdir", type=str, required=False,
                        default=Defaults.conf_dir, help="ceph configuration directory")
    parser.add_argument("-u", "--user", type=str, required=False,
                        default=Defaults.user, help="Ceph user to use for the keyring")
    parser.add_argument("-f", "--format", type=str, required=False,
                        choices=Choices.output_format,
                        default=Defaults.output_format, 
                        help="output file format")
    return parser.parse_args()

def main():

    ok, _err = ready()
    if not ok:
        abort(_err)

    conf = get_config(os.path.join(args.confdir, 'ceph.conf'))
    keyring = find_keyring()
    if not keyring:
        abort("Can't find a keyring to use")

    print("Fetching state of the local Ceph cluster")
    ceph_out, error = send_command('ceph -s -f json')
    if error:
        abort("ceph command failed - can't determine state")
    
    ceph_version_text, error = send_command('ceph --version')
    if error:
        abort("unable to fetch ceph version!")

    ceph_state = json.loads(ceph_out)

    settings = dict()

    settings['fsid'] = ceph_state['fsid']
    settings['version'] = str(ceph_version_text).split()[2].split('-')[0]  # just want a V.R.M for the version id
    settings['secret'] = keyring
    settings['mons'] = [mon['public_addr'].split(':')[0] for mon in ceph_state['monmap']['mons']]
    settings['mgr'] = ceph_state['mgrmap']['active_addr'].split(':')[0]
   
    dashboard_url = ceph_state['mgrmap']['services'].get('dashboard', None)
    if dashboard_url:
        settings['dashboard'] = dashboard_url
   
    if ceph_state.get('servicemap', None):
        if ceph_state['servicemap']['services'].get('rgw', None):
            daemons = ceph_state['servicemap']['services']['rgw']['daemons']
            rgws = list()
            for rgw in daemons:
                if rgw == "summary":
                    continue
                frontend_str = daemons[rgw]['metadata']['frontend_config#0']
                rgws.append(frontend_str.split('port=')[1].split()[0])
            settings['rgws'] = rgws
        
    dump(settings)


if __name__ == '__main__':
    args = parse_args()
    main()

File: go/ceph-export.go
package main

//
// export the ceph configuration
// code arbitrarily navigates through ceph -s json, instead of fully declaring
// a struct to define the whole of ceph -s output. This could be done later if
// desired
//

import (
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"os/exec"
	"os/user"
	"strconv"
	"strings"

	"gopkg.in/ini.v1"
	"gopkg.in/yaml.v2"
)

// keyringFile is the filename pattern for a keyring
const keyringFile = "ceph.client.%s.keyring"

var defaults = map[string]string{
	"outFile":    "~/rhcs-export",
	"confDir":    "/etc/ceph",
	"fileFormat": "json",
	"userName":   "admin",
}

// Runtime settings
type runtimeSettings struct {
	outFile    string
	confDir    string
	fileFormat string
	userName   string
}

// exported ceph configuration metadata
type cephMetaData struct {
	DashboardURL  string   `json:"dashboard_url" yaml:"dashboard_url"`
	Fsid          string   `json:"fsid" yaml:"fsid"`
	Secret        string   `json:"secret" yaml:"secret"`
	Mgr           string   `json:"mgr" yaml:"mgr"`
	Mgrstandby    []string `json:"mgr_standby" yaml:"mgr_standby"`
	Mons          []string `json:"mons" yaml:"mons"`
	PrometheusURL string   `json:"prometheus_url" yaml:"prometheus_url"`
	Rgws          []string `json:"rgws" yaml:"rgws"`
	Version       string   `json:"version" yaml:"version"`
}

func isDir(filePath string) bool {
	info, err := os.Stat(filePath)
	if os.IsNotExist(err) {
		return false
	}
	return info.IsDir()
}

func isFile(filePath string) bool {
	info, err := os.Stat(filePath)
	if os.IsNotExist(err) {
		return false
	}
	return !info.IsDir()
}

// look for string in a given slice
func hasString(item string, iterable []string) bool {
	for _, value := range iterable {
		if item == value {
			return true
		}
	}
	return false
}

// check if the environment is suitable for the export
func ready(settings *runtimeSettings) (bool, error) {

	keyring := fmt.Sprintf(keyringFile, settings.userName)
	keyringStore := settings.confDir + "/keyring-store/keyring"

	if !isDir(settings.confDir) {
		return false, errors.New("Directory '" + settings.confDir + "' not found")
	}

	if !isFile(settings.confDir + "/ceph.conf") {
		return false, errors.New("ceph configuration file missing from " + settings.confDir)
	}

	if !isFile(settings.confDir+"/"+keyring) && !isFile(keyringStore) {
		return false, errors.New("missing keyring/keyring store")
	}

	_, err := sendCommand("type ceph")
	if err != nil {
		return false, errors.New("ceph command is unavailable")
	}

	return true, nil
}

// exit the program with an error message
func abort(message string) {
	fmt.Printf("Unable to continue: %s\n", message)
	os.Exit(4)
}

// send a command to the OS, and return the response to the caller
func sendCommand(commandString string) (string, error) {
	args := strings.Split(commandString, " ")

	cmd := exec.Command(args[0], args[1:]...)
	out, err := cmd.Output()
	if err != nil {
		return "", errors.New("error running command")
	}
	return string(out), nil
}

// find the keyring for the given user and return its key
func fetchKeyring(userName string, confDir string) string {

	var keyFile string

	keyFilePath := confDir + "/" + fmt.Sprintf(keyringFile, userName)
	keyStore := confDir + "/keyring-store/keyring"
	if isFile(keyFilePath) {
		keyFile = keyFilePath
	} else if isFile(keyStore) {
		keyFile = keyStore
	}

	// what if keyFile is not set i.e. still empty?

	conf, err := getConfig(keyFile)
	if err != nil {
		return ""
	}
	keySection := conf.Section("client." + userName)
	key, err := keySection.GetKey("key")
	if err != nil {
		return ""
	}

	return key.String()
}

// simplistic hstname check -if ir starts with a number, it's an IP address!
func isIP(hostName string) bool {
	char1 := string(hostName[0])
	_, err := strconv.ParseInt(char1, 10, 8)
	if err != nil {
		return false
	}
	return true
}

// Read a ceph confg (ini) format
func getConfig(confFileName string) (*ini.File, error) {
	cfg, err := ini.Load(confFileName)
	if err != nil {
		return cfg, errors.New("Unable to load the config file")
	}
	return cfg, nil
}

// export to a file
func writeFile(output []byte, settings *runtimeSettings) {
	if strings.HasPrefix(settings.outFile, "~") {
		usr, _ := user.Current()
		settings.outFile = strings.Replace(settings.outFile, "~", usr.HomeDir, 1)
	}
	fileName := settings.outFile + "." + settings.fileFormat

	err := ioutil.WriteFile(fileName, output, 0644)
	if err != nil {
		fmt.Println(err)
		abort("Failed to write the file ")
	} else {
		fmt.Println("\nMetadata written to " + fileName)
	}
}

// dump to json
func toJSON(content *cephMetaData) []byte {

	out, err := json.MarshalIndent(content, "", "    ")
	if err != nil {
		abort("Export to json failed")
	}
	return out
}

// dump to yaml
func toYAML(content *cephMetaData) []byte {

	out := []byte("---\n")
	yaml, err := yaml.Marshal(content)
	if err != nil {
		abort("Export to yaml failed")
	}
	return append(out, yaml...)
}

// write ceph facts to a file
func exportMetadata(content *cephMetaData, settings *runtimeSettings) error {

	// fmt.Println(*content)
	// fmt.Println(*settings)
	switch settings.fileFormat {
	case "json":
		out := toJSON(content)
		writeFile(out, settings)
		// fmt.Println(string(out))
	case "yaml":
		out := toYAML(content)
		writeFile(out, settings)
	}

	return nil
}

func main() {

	var enabledModules []string
	var exportData cephMetaData

	// Defaults for the command line args
	outFile := flag.String("output", defaults["outFile"], "output file name")
	confDir := flag.String("confdir", defaults["confDir"], "Ceph configuration directory")
	fileFormat := flag.String("format", defaults["fileFormat"], "output file format")
	userName := flag.String("user", defaults["userName"], "user keyring")

	flag.Parse()
	settings := runtimeSettings{
		outFile:    *outFile,
		confDir:    *confDir,
		fileFormat: *fileFormat,
		userName:   *userName,
	}

	fmt.Print("\nChecking environment......")
	ok, err := ready(&settings)
	if !ok {
		fmt.Print("FAILED\n")
		abort(err.Error())
	} else {
		fmt.Print("PASSED\n")
	}

	key := fetchKeyring(settings.userName, settings.confDir)
	if key == "" {
		abort("Unable to load a key for the '" + *userName + "' user")
	}

	fmt.Print("Querying ceph state.......")
	cephStatusStr, err := sendCommand("ceph -s -f json")
	if err != nil {
		fmt.Print("FAILED")
		abort("Unable to gather status from ceph with 'ceph -s' command")
	} else {
		fmt.Print("OK\n")
	}

	bytes := []byte(cephStatusStr)
	var cephStatus map[string]interface{}
	err = json.Unmarshal(bytes, &cephStatus)
	if err != nil {
		abort("Unable to parse the json output from Ceph!")
	}

	fmt.Print("Checking ceph version.....")
	cephVersOutput, err := sendCommand("ceph --version")
	if err != nil {
		fmt.Print("FAILED\n")
		abort("failed trying to extract ceph version from the system")
	}

	// the first 13 chars are 'ceph version ', so let's skip those!
	exportData.Version = strings.Split(cephVersOutput[13:], "-")[0]
	if !strings.HasPrefix(exportData.Version, "14") {
		abort("Export utility only supported on Nautilus clusters")
	} else {
		fmt.Print("PASSED\n")
	}

	for idx, k := range cephStatus {

		switch idx {
		case "monmap":
			// fmt.Println("Processing monmap")
			for midx, mval := range k.(map[string]interface{}) {
				switch midx {
				case "mons":
					// fmt.Println("processing mons")
					for _, monData := range mval.([]interface{}) {
						monIP := monData.(map[string]interface{})["addr"]
						exportData.Mons = append(exportData.Mons, monIP.(string))
					}
				}
			}
		case "mgrmap":
			// fmt.Println("Processing mgrmap")
			for mgrKey, mgrVal := range k.(map[string]interface{}) {

				switch mgrKey {
				case "active_addr":
					exportData.Mgr = strings.Split(mgrVal.(string), ":")[0]
				case "standbys":
					for _, stdbyData := range mgrVal.([]interface{}) {
						s := stdbyData.(map[string]interface{})
						mgrName := s["name"].(string)
						if !isIP(mgrName) {
							ip, err := net.LookupHost(mgrName)
							if err == nil {
								mgrName = ip[0]
							}
						}
						exportData.Mgrstandby = append(exportData.Mgrstandby, mgrName)
					}
				case "modules":
					for _, mod := range mgrVal.([]interface{}) {
						enabledModules = append(enabledModules, mod.(string))
					}
				case "services":
					for svcName, svcURL := range mgrVal.(map[string]interface{}) {
						switch svcName {
						case "dashboard":
							exportData.DashboardURL = svcURL.(string)
						case "prometheus":
							exportData.PrometheusURL = svcURL.(string)
						}
					}
				}
			}
		case "servicemap":
			svcMap := k.(map[string]interface{})
			svcs := svcMap["services"].(map[string]interface{})

			if rgw, ok := svcs["rgw"]; ok {
				rgwDaemons := rgw.(map[string]interface{})["daemons"]
				for rgwKey, rgwData := range rgwDaemons.(map[string]interface{}) {
					if rgwKey == "summary" {
						continue
					}
					rgwMeta := rgwData.(map[string]interface{})["metadata"]
					frontEnd := rgwMeta.(map[string]interface{})["frontend_config#0"].(string)
					fendSettings := strings.Split(frontEnd, " ")
					for _, item := range fendSettings {
						parms := strings.Split(item, "=")
						if strings.HasSuffix(parms[0], "port") {
							exportData.Rgws = append(exportData.Rgws, parms[1])
						}
					}
				}
			}

		}

	}

	if !hasString("prometheus", enabledModules) {
		abort("Prometheus module must be enabled, prior to configuration export")
	}
	fmt.Println("Active mgr module check...PASSED")

	exportData.Secret = key
	exportData.Fsid = cephStatus["fsid"].(string)

	exportMetadata(&exportData, &settings)
}
